### ISSUE TITLE
SSL error

### ISSUE THREAD

**Issue by ihxkjggg** [ZEISS-PiWeb/PiWeb-Api#95]
How do i use Python to access piwed database? I always prompt SSL error

urllib.error.URLError: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1108)>

**Comment by czthiele** [comment]
Hello @ihxkjggg,

this seems to be a problem with the certificate used by the PiWeb Server for SSL/HTTPS. Python cannot check if the certificate is valid. Make sure that the root certificate used by PiWeb Server is trusted on your own pc an accessible by Python (you may need to add it to the trusted roots used by Python). Can you reach the PiWeb Server with your browser without any certificate issues?

**Comment by ihxkjggg** [comment]
Hello@czthiele,
After I input the web address in my computer, a Windows window will pop up for verification. After I input the account password, I can get the JSON data.
However, when using Python to request the URL, it will prompt SSL error.
![ssl](https://user-images.githubusercontent.com/69021797/118778436-ad8d1980-b8bc-11eb-922e-b13d021d62bf.jpg)

**Comment by ihxkjggg** [comment]
Hello @czthiele 
After I input the web address in my computer, a Windows window will pop up for verification（It's like the picture below). After I input the account password, I can get the JSON data.
However, when using Python to request the URL, it will prompt SSL error.
![ssl](https://user-images.githubusercontent.com/69021797/118779446-9f8bc880-b8bd-11eb-818c-8051a36ad393.jpg)

**Comment by czthiele** [comment]
Hello @ihxkjggg,
which version of PiWeb Server is running and which authentication type is used?

**Comment by ihxkjggg** [comment]
Thank you very much. Because I can only access it under the company's LAN, now I can't view the corresponding piweb information after work. I can only give you a clear reply tomorrow. I hope it can be solved with your help.

**Comment by ihxkjggg** [comment]
@czthiele 
Thank you very much. Because I can only access it under the company's LAN, now I can't view the corresponding piweb information after work. I can only give you a clear reply tomorrow. I hope it can be solved with your help.

**Comment by ihxkjggg** [comment]
@czthiele 
Hello czthiele ,I got the piweb information, and manually input the URL to capture the package, the following is the details, please help to see, thank you
[piweb.pdf](https://github.com/ZEISS-PiWeb/PiWeb-Api/files/6512407/piweb.pdf)

**Comment by ihxkjggg** [comment]
Hello, @czthiele 
can you help me solve this problem? Thank you very much.

**Comment by czthiele** [comment]
Hello @ihxkjggg , please excuse the waiting time. I will further analyze this issue today.

**Comment by czthiele** [comment]
Hello @ihxkjggg,

one thing you could try is to disable the SSL verfification to test if this is causing the problem. In your requests.get(...) simply set verify=False. If you can reach the server with this configuration then the problem seems to be the SSL certificate, or that Python does not trust the root certificate of your PiWeb Server certificate. This is something not related to PiWeb itself but to Python, and since I am no Python specialist I can only give you some helpful links on how to add certificates to Python:
https://incognitjoe.github.io/adding-certs-to-requests.html
https://appdividend.com/2020/06/19/python-certifi-example-how-to-use-ssl-certificate-in-python/

Another hint: try changing the Connection header from 'close' to 'keep-alive'. And don't forget to enter your valid credentials in the requests 'Domain\\User' and 'password' (but i guess you just removed them to post your code example here).

I guess this is all I can do for now since I cannot help with Python issues. Please let me know if it somehow helped.

**Comment by ihxkjggg** [comment]
hello, @czthiele ,
Thank you for the method. I tried it, but I didn't solve it. I'm also trying other methods. If you have successful methods in the future, I hope you can share them. Thank you very much.

**Comment by czbod** [comment]
hello, is there meanwhile a solution for this problem?

**Comment by czthiele** [comment]
Hello @czbod, 
I recently tried certificate authentication using Python and had a similar issue (caused by Python and not PiWeb itself). As stated here earlier, the main problem was the SSL certficate used by PiWeb Server for HTTPS. Since Python seems to not use the users windows certificate storage we need to add the SSL certificate CA to Pythons trusted certificate authorities. 
For this I used the [certifi package](https://pypi.org/project/certifi/) which adds common root certificates and gives us the possibility to add our own trusted certificates. certifi.where() will give you the path to the certificate-file (cacert.pem). You simply need to append the certificate authority which issued your SSL certificate (or certificate authorities if there are intermediate certificates in the chain). A trick to get the CA certificate in the .pem format is to export it in base64 encoding and then simply change the file ending to pem. Now append the content of this file to the cacert.pem of certifi.
Now Python should be able to get the local issuer certificate since it is inside the cacert.pem.

I hope that this will help.
